### PR TITLE
pageserver: return proper status code for heatmap_upload errors

### LINKED ISSUE #9574: pageserver: noisy heatmap upload error during pageserver deploy: ` request{method=POST path=/v1/tenant/TENANT_ID/heatmap_upload ...}: Error processing HTTP request: InternalServerError(Tenant TENANT_ID is not active`
# Problem

During pageserver deploys, heatmap upload logs noisy errors

```
2024-10-29T16:20:30.253228Z ERROR request{method=POST path=/v1/tenant/TENANT_ID/heatmap_upload request_id=091ea8ff-1777-4bbd-ba41-3c74190ab0f5}: Error processing HTTP request: InternalServerError(Tenant TENANT_ID is not active
```

<details>
```
2024-10-29T16:20:30.253228Z ERROR request{method=POST path=/v1/tenant/TENANT_ID/heatmap_upload request_id=091ea8ff-1777-4bbd-ba41-3c74190ab0f5}: Error processing HTTP request: InternalServerError(Tenant TENANT_ID is not active
Stack backtrace:
   0: <T as core::convert::Into<U>>::into
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/core/src/convert/mod.rs:759:9
      <pageserver::tenant::secondary::heatmap_uploader::HeatmapUploader as pageserver::tenant::secondary::scheduler::JobGenerator<pageserver::tenant::secondary::heatmap_uploader::UploadPending,pageserver::tenant::secondary::heatmap_uploader::WriteInProgress,pageserver::tenant::secondary::heatmap_uploader::WriteComplete,pageserver::tenant::secondary::UploadCommand>>::on_command
             at /home/nonroot/pageserver/src/tenant/secondary/heatmap_uploader.rs:293:68
   1: pageserver::tenant::secondary::scheduler::TenantBackgroundJobs<G,PJ,RJ,C,CMD>::handle_command
             at /home/nonroot/pageserver/src/tenant/secondary/scheduler.rs:327:25
   2: pageserver::tenant::secondary::scheduler::TenantBackgroundJobs<G,PJ,RJ,C,CMD>::run::{{closure}}
             at /home/nonroot/pageserver/src/tenant/secondary/scheduler.rs:238:25
   3: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.40/src/instrument.rs:321:9
   4: pageserver::tenant::secondary::heatmap_uploader::heatmap_uploader_task::{{closure}}
             at /home/nonroot/pageserver/src/tenant/secondary/heatmap_uploader.rs:57:10
      pageserver::tenant::secondary::spawn_tasks::{{closure}}
             at /home/nonroot/pageserver/src/tenant/secondary.rs:389:14
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::future::future::Future>::poll
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/core/src/panic/unwind_safe.rs:297:9
   5: <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll::{{closure}}
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.30/src/future/future/catch_unwind.rs:36:42
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/core/src/panic/unwind_safe.rs:272:9
      std::panicking::try::do_call
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/std/src/panicking.rs:554:40
      std::panicking::try
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/std/src/panicking.rs:518:19
      std::panic::catch_unwind
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/std/src/panic.rs:345:14
      <futures_util::future::future::catch_unwind::CatchUnwind<Fut> as core::future::future::Future>::poll
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/futures-util-0.3.30/src/future/future/catch_unwind.rs:36:9
      pageserver::task_mgr::exit_on_panic_or_error::{{closure}}
             at /home/nonroot/pageserver/src/task_mgr.rs:522:58
   6: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.1/src/runtime/task/core.rs:328:17
      tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.1/src/loom/std/unsafe_cell.rs:16:9
      tokio::runtime::task::core::Core<T,S>::poll
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.1/src/runtime/task/core.rs:317:30
   7: tokio::runtime::task::harness::poll_future::{{closure}}
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.1/src/runtime/task/harness.rs:485:19
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/core/src/panic/unwind_safe.rs:272:9
      std::panicking::try::do_call
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/std/src/panicking.rs:554:40
      std::panicking::try
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/std/src/panicking.rs:518:19
      std::panic::catch_unwind
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/std/src/panic.rs:345:14
      tokio::runtime::task::harness::poll_future
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.1/src/runtime/task/harness.rs:473:18
      tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.1/src/runtime/task/harness.rs:208:27
      tokio::runtime::task::harness::Harness<T,S>::poll
             at /home/nonroot/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.38.1/src/runtime/task/harness.rs:153:15
   8: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
   9: tokio::runtime::scheduler::multi_thread::worker::Context::run
  10: tokio::runtime::context::scoped::Scoped<T>::set
  11: tokio::runtime::context::runtime::enter_runtime
  12: tokio::runtime::scheduler::multi_thread::worker::run
  13: tokio::runtime::task::core::Core<T,S>::poll
  14: tokio::runtime::task::harness::Harness<T,S>::poll
  15: tokio::runtime::blocking::pool::Inner::run
  16: std::sys::backtrace::__rust_begin_short_backtrace
  17: core::ops::function::FnOnce::call_once{{vtable.shim}}
  18: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/alloc/src/boxed.rs:2231:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/alloc/src/boxed.rs:2231:9
      std::sys::pal::unix::thread::Thread::new::thread_start
             at /rustc/f6e511eec7342f59a25f7c0534f1dbea00d01b14/library/std/src/sys/pal/unix/thread.rs:105:17
  19: <unknown>
  20: <unknown>)
```

</details>

# Solution

The current architecture doesn't allow elimination of this toctu situation.
So, introduce error types / variants to bubble up cancellation reason.

### PR DESCRIPTION
**Pull request by erikgrinaker**
## Problem

During deploys, we see a lot of 500 errors due to heapmap uploads for inactive tenants. These should be 503s instead.

Resolves #9574.

## Summary of changes

Make the secondary tenant scheduler use `ApiError` rather than `anyhow::Error`, to propagate the tenant error and convert it to an appropriate status code.

### CODE REVIEW

File: pageserver/src/tenant/secondary.rs [hunk 1]
```diff
@@ -66,7 +66,7 @@ struct CommandRequest<T> {
 }
 
 struct CommandResponse {
-    result: anyhow::Result<()>,
+    result: Result<(), ApiError>,
```
Comment (erikgrinaker) [inline pageserver/src/tenant/secondary.rs:69]: Alternatively, we could make the scheduler generic over the error type, but this seems fine for now.

File: pageserver/src/tenant/secondary.rs [hunk 2]
```diff
@@ -294,20 +294,18 @@ impl SecondaryController {
                 response_tx,
             })
             .await
-            .map_err(|_| anyhow::anyhow!("Receiver shut down"))?;
+            .map_err(|_| ApiError::ShuttingDown)?;
 
-        let response = response_rx
-            .await
-            .map_err(|_| anyhow::anyhow!("Request dropped"))?;
+        let response = response_rx.await.map_err(|_| ApiError::ShuttingDown)?;
```
Comment (erikgrinaker) [inline pageserver/src/tenant/secondary.rs:299]: Is `ShuttingDown` appropriate here?

**Review by jcsp — APPROVED** [review]
LGTM.

In most of the code we have internal error types that get converted to ApiError at the interface (routes.rs).  Using ApiError for `on_command` path is intuitively reasonable as we invoke that via APIs, although using it paths like `upload_tenant` is a little bit odd as these run in the background & we could end up logging API errors for things that weren't called from an API.

Nevertheless, this is a net improvement compared with munging things into anyhow::Error, and it gets the job done

File: pageserver/src/tenant/secondary.rs (hunk 2 above)
Comment (jcsp) [reply pageserver/src/tenant/secondary.rs:299]: Seems reasonable to me

**Comment by erikgrinaker** [comment]
> In most of the code we have internal error types that get converted to ApiError at the interface (routes.rs). Using ApiError for `on_command` path is intuitively reasonable as we invoke that via APIs, although using it paths like `upload_tenant` is a little bit odd as these run in the background & we could end up logging API errors for things that weren't called from an API.

Yeah, I agree. Let me see how much hassle it would be to add an intermediate error type here.

**Comment by erikgrinaker** [comment]
Added `GetTenantError::ShardNotFound` and `SecondaryTenantError`, PTAL.

**Review by jcsp — APPROVED** [review]

**Pull request merged**